packages/loader/container-loader/src/serializedStateManager.ts

```diff
@@ -0,0 +1,201 @@
+		>,
+		private readonly _offlineLoadEnabled: boolean,
+	) {
+		this.mc = createChildMonitoringContext({
+			logger: subLogger,
+			namespace: "serializedStateManager",
+		});
+		if (pendingLocalState !== undefined && _offlineLoadEnabled) {
+			this.snapshot = {
+				tree: pendingLocalState.baseSnapshot,
+				blobs: pendingLocalState.snapshotBlobs,
+			};
+		}
+	}
+
+	public get offlineLoadEnabled(): boolean {
+		return this._offlineLoadEnabled;
+	}
+
+	public addSavedOp(message: ISequencedDocumentMessage) {
+		if (this.offlineLoadEnabled) {
+			this.savedOps.push(message);
+		}
+	}
+
+	private async getVersion(version: string | null): Promise<IVersion | undefined> {
+		const versions = await this.storageAdapter.getVersions(version, 1);
+		return versions[0];
+	}
+
+	public async fetchSnapshot(
+		specifiedVersion: string | undefined,
+		supportGetSnapshotApi: boolean | undefined,
+	) {
+		const { snapshot, version } =
+			this.pendingLocalState === undefined
+				? await this.fetchSnapshotCore(specifiedVersion, supportGetSnapshotApi)
+				: { snapshot: this.pendingLocalState.baseSnapshot, version: undefined };
+		const snapshotTree: ISnapshotTree | undefined = isInstanceOfISnapshot(snapshot)
+			? snapshot.snapshotTree
+			: snapshot;
+		assert(snapshotTree !== undefined, "Snapshot should exist");
+		// non-interactive clients will not have any pending state we want to save
+		if (this.offlineLoadEnabled) {
+			const blobs = await getBlobContentsFromTree(snapshotTree, this.storageAdapter);
+			this.snapshot = { tree: snapshotTree, blobs };
+		}
+		return { snapshotTree, version };
+	}
+
+	private async fetchSnapshotCore(
+		specifiedVersion: string | undefined,
+		supportGetSnapshotApi: boolean | undefined,
+	): Promise<{ snapshot?: ISnapshot | ISnapshotTree; version?: IVersion }> {
+		if (
+			this.mc.config.getBoolean("Fluid.Container.UseLoadingGroupIdForSnapshotFetch") ===
+				true &&
+			supportGetSnapshotApi === true
+		) {
+			const snapshot =
+				(await this.storageAdapter.getSnapshot?.({
+					versionId: specifiedVersion,
+				})) ?? undefined;
+			const version: IVersion = {
+				id: snapshot?.snapshotTree.id ?? "",
+				treeId: snapshot?.snapshotTree.id ?? "",
+			};
+
+			if (snapshot === undefined && specifiedVersion !== undefined) {
+				this.mc.logger.sendErrorEvent({
+					eventName: "getSnapshotTreeFailed",
+					id: version.id,
+				});
+			}
+			return { snapshot, version };
+		}
+		return this.fetchSnapshotTree(specifiedVersion);
+	}
+
+	/**
+	 * Get the most recent snapshot, or a specific version.
+	 * @param specifiedVersion - The specific version of the snapshot to retrieve
+	 * @returns The snapshot requested, or the latest snapshot if no version was specified, plus version ID
+	 */
+	private async fetchSnapshotTree(
+		specifiedVersion: string | undefined,
+	): Promise<{ snapshot?: ISnapshotTree; version?: IVersion | undefined }> {
+		const version = await this.getVersion(specifiedVersion ?? null);
```

Comment:
Why does getVersion take in null if there's no version specified, instead of undefined? This is more of a general curiosity, since I don't think you set these parameter types, but I thought I'd ask here. Maybe @anthony-murphy knows?